File: cpp/base/src/vsdk/toolkit/skin/MeshSurface.cpp
#include "vsdk/toolkit/java/util/ArrayList.txx"
#include "vsdk/toolkit/common/statistics/Statistics.h"
#include "vsdk/toolkit/skin/MeshSurface.h"

int MeshSurface::nextSurfaceId = 0;

/**
Indicates on whether or not, and if so, which, colors are given when creating
a new surface
*/
MaterialColorFlags MeshSurface::colorFlags = NO_COLORS;

/**
This routine creates a MeshSurface with given inMaterial, positions
*/
MeshSurface::MeshSurface(
    char *inObjectName,
    Material *inMaterial,
    java::ArrayList<Vector3D *> *inPoints,
    java::ArrayList<Vector3D *> *inNormals,
    const java::ArrayList<Vector3D *> * /*texCoords*/,
    java::ArrayList<Vertex *> *inVertices,
    java::ArrayList<Patch *> *inFaces,
    MaterialColorFlags inFlags)
{
    Statistics::instance().reader.numberOfSurfaces++;

    id = nextGeometryId;
    nextGeometryId++;
    objectName = inObjectName;
    meshId = nextSurfaceId++;
    className = GeometryClassId::SURFACE_MESH;
    isDuplicate = false;

    material = inMaterial;
    positions = inPoints;
    normals = inNormals;
    vertices = inVertices;
    faces = inFaces;

    colorFlags = inFlags;

    // If colorFlags == VERTEX_COLORS< the inVertices are assumed to contain
    // the sum of the colors as used in each patch sharing the vertex
    if ( colorFlags == VERTEX_COLORS ) {
        for ( int i = 0; vertices != nullptr && i < vertices->size(); i++ ) {
            MeshSurface::normalizeVertexColor(vertices->get(i));
        }
    }

    // Compute vertex colors
    if ( colorFlags != VERTEX_COLORS ) {
        for ( int i = 0; vertices != nullptr && i < vertices->size(); i++ ) {
            vertices->get(i)->computeColor();
        }
    }

    colorFlags = NO_COLORS;

    patchListBounds(faces, &boundingBox);

    // Enlarge bounding box a tiny bit for more conservative bounding box culling
    boundingBox.enlargeTinyBit();
    bounded = true;
    shaftCullGeometry = false;
    radianceData = nullptr;
    itemCount = 0;
    omit = false;
}

MeshSurface::~MeshSurface() {
    if ( objectName ) {
        delete[] objectName;
    }

    if ( positions != nullptr) {
        for ( int i = 0; i < positions->size(); i++ ) {
            delete positions->get(i);
        }
        delete positions;
    }

    if ( normals != nullptr ) {
        for ( int i = 0; i < normals->size(); i++ ) {
            delete normals->get(i);
        }
        delete normals;
    }

    if ( vertices != nullptr ) {
        for ( int i = 0; i < vertices->size(); i++ ) {
            delete vertices->get(i);
        }
        delete vertices;
    }

    if ( faces != nullptr ) {
        for ( int i = 0; i < faces->size(); i++ ) {
            delete faces->get(i);
        }
        delete faces;
    }
}

void
MeshSurface::normalizeVertexColor(Vertex *vertex) {
    long numberOfPatches = 0;

    if ( vertex->patches != nullptr ) {
        numberOfPatches = vertex->patches->size();
    }

    if ( numberOfPatches > 0 ) {
        const double invPatchCount = 1.0 / static_cast<double>(numberOfPatches);
        vertex->color = ColorRgb(
            vertex->color.getR() * invPatchCount,
            vertex->color.getG() * invPatchCount,
            vertex->color.getB() * invPatchCount);
    }
}

/**
DiscretizationIntersect returns nullptr is the ray doesn't hit the discretization
of the object. If the ray hits the object, a hit record is returned containing
information about the intersection point. See geometry.h for more explanation
*/
RayHit *
MeshSurface::discretizationIntersect(
    Ray *ray,
    float minimumDistance,
    float *maximumDistance,
    int hitFlags,
    RayHit *hitStore) const
{
    return patchListIntersect(faces, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
}

File: typescript/base/src/main/vsdk/toolkit/skin/MeshSurface.ts
import { Ray } from "../common/linealAlgebra/Ray";
import { Vector3D } from "../common/linealAlgebra/Vector3D";
import { Statistics } from "../common/statistics/Statistics";
import { Material } from "../material/Material";
import { Geometry } from "./Geometry";
import { GeometryClassId } from "./GeometryClassId";
import { MaterialColorFlags } from "../material/MaterialColorFlags";
import type { Patch } from "../environment/geometry/elements/Patch";
import type { RayHit } from "../environment/geometry/elements/RayHit";
import { Vertex } from "../environment/geometry/elements/Vertex";

export class MeshSurface extends Geometry {
  private static nextSurfaceId = 0;
  public static colorFlags = MaterialColorFlags.NO_COLORS;

  public meshId: number;
  public objectName: string | null;
  public vertices: Vertex[] | null;
  public positions: Vector3D[] | null;
  public normals: Vector3D[] | null;
  public faces: Patch[] | null;
  public material: Material;

  public constructor(
    inObjectName: string,
    inMaterial: Material,
    inPoints: Vector3D[] | null,
    inNormals: Vector3D[] | null,
    texCoords: Vector3D[] | null,
    inVertices: Vertex[] | null,
    inFaces: Patch[] | null,
    inFlags: number
  ) {
    super();

    Statistics.instance().reader.numberOfSurfaces++;

    this.id = Geometry.nextGeometryId;
    Geometry.nextGeometryId++;
    this.objectName = inObjectName;
    this.meshId = MeshSurface.nextSurfaceId++;
    this.className = GeometryClassId.SURFACE_MESH;
    this.isDuplicate = false;

    this.material = inMaterial;
    this.positions = inPoints;
    this.normals = inNormals;
    this.vertices = inVertices;
    this.faces = inFaces;

    MeshSurface.colorFlags = inFlags;

    if (MeshSurface.colorFlags === MaterialColorFlags.VERTEX_COLORS) {
      for (let i = 0; this.vertices !== null && i < this.vertices.length; i++) {
        MeshSurface.normalizeVertexColor(this.vertices[i]!);
      }
    }

    if (MeshSurface.colorFlags !== MaterialColorFlags.VERTEX_COLORS) {
      for (let i = 0; this.vertices !== null && i < this.vertices.length; i++) {
        this.vertices[i]!.computeColor();
      }
    }

    MeshSurface.colorFlags = MaterialColorFlags.NO_COLORS;

    Geometry.patchListBounds(this.faces, this.boundingBox);
    this.boundingBox.enlargeTinyBit();
    this.bounded = true;
    this.shaftCullGeometry = false;
    this.radianceData = null;
    this.itemCount = 0;
    this.omit = false;

    void texCoords;
  }

  public override destroy(): void {
    this.objectName = null;

    if (this.positions !== null) {
      this.positions.length = 0;
      this.positions = null;
    }

    if (this.normals !== null) {
      this.normals.length = 0;
      this.normals = null;
    }

    if (this.vertices !== null) {
      for (let i = 0; i < this.vertices.length; i++) {
        this.vertices[i]!.destroy();
      }
      this.vertices.length = 0;
      this.vertices = null;
    }

    if (this.faces !== null) {
      for (let i = 0; i < this.faces.length; i++) {
        this.faces[i]!.destroy();
      }
      this.faces.length = 0;
      this.faces = null;
    }

    super.destroy();
  }

  private static normalizeVertexColor(vertex: Vertex): void {
    let numberOfPatches = 0;
    if (vertex.patches !== null) {
      numberOfPatches = vertex.patches.length;
    }

    if (numberOfPatches > 0) {
      vertex.color.r /= numberOfPatches;
      vertex.color.g /= numberOfPatches;
      vertex.color.b /= numberOfPatches;
    }
  }

  public override discretizationIntersect(
    ray: Ray,
    minimumDistance: number,
    maximumDistance: number[],
    hitFlags: number,
    hitStore: RayHit | null
  ): RayHit | null {
    return Geometry.patchListIntersect(this.faces, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
  }
}
